lib/SourceMapDevToolPlugin.js

```diff
@@ -163,233 +161,279 @@ class SourceMapDevToolPlugin {
-								throw new Error(
-									"SourceMapDevToolPlugin: append can't be false when no filename is provided"
+
+							// find modules with conflicting source names
+							for (let idx = 0; idx < allModules.length; idx++) {
+								const module = allModules[idx];
+								let sourceName = moduleToSourceNameMapping.get(module);
+								let hasName = conflictDetectionSet.has(sourceName);
+								if (!hasName) {
+									conflictDetectionSet.add(sourceName);
+									continue;
+								}
+
+								// try the fallback name first
+								sourceName = ModuleFilenameHelpers.createFilename(
+									module,
+									{
+										moduleFilenameTemplate: fallbackModuleFilenameTemplate,
+										namespace: namespace
+									},
+									{
+										requestShortener,
+										chunkGraph
+									}
 								);
+								hasName = usedNamesSet.has(sourceName);
+								if (!hasName) {
+									moduleToSourceNameMapping.set(module, sourceName);
+									usedNamesSet.add(sourceName);
+									continue;
+								}
+
+								// elsewise just append stars until we have a valid name
+								while (hasName) {
+									sourceName += "*";
+									hasName = usedNamesSet.has(sourceName);
+								}
+								moduleToSourceNameMapping.set(module, sourceName);
+								usedNamesSet.add(sourceName);
 							}
-							/**
-							 * Add source map as data url to asset
-							 */
-							const asset = new ConcatSource(
-								new RawSource(source),
-								currentSourceMappingURLComment
-									.replace(/\[map\]/g, () => sourceMapString)
-									.replace(
-										/\[url\]/g,
-										() =>
-											`data:application/json;charset=utf-8;base64,${Buffer.from(
-												sourceMapString,
-												"utf-8"
-											).toString("base64")}`
-									)
+
+							let taskIndex = 0;
+
+							asyncLib.each(
+								tasks,
+								(task, done) => {
+									reportProgress(
+										0.5 + (0.5 * taskIndex++) / tasks.length,
+										task.file,
+										"attach SourceMap"
+									);
+									const assets = Object.create(null);
+									const file = task.file;
+									const chunk = fileToChunk.get(file);
+									const asset = task.asset;
+									const sourceMap = task.sourceMap;
+									const source = task.source;
+									const modules = task.modules;
+									const moduleFilenames = modules.map(m =>
+										moduleToSourceNameMapping.get(m)
+									);
+									sourceMap.sources = moduleFilenames;
+									if (options.noSources) {
+										sourceMap.sourcesContent = undefined;
+									}
+									sourceMap.sourceRoot = options.sourceRoot || "";
+									sourceMap.file = file;
+
+									const cacheIdent = `${compilation.compilerPath}/SourceMapDevToolPlugin/${file}`;
+									const cacheETag = createHash("md4");
+									asset.updateHash(cacheETag);
+
+									compilation.cache.store(
+										cacheIdent,
+										cacheETag.digest("hex").toString(),
+										assets,
```

Comment:
This is store to early. `assets` is not written yet. Store it later after `assets` is done. e.g. line 428